### PR TITLE
coprocessor/dag/expr/builtin_time:  implement  period_add & period_diff

### PR DESCRIPTION
**Pull request by Seact**
Signed-off-by: Seact seact_l@outlook.com

## What have you changed? (mandatory)

This PR implements the scalar function PeriodAdd and PeriodDiff

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- New feature (change which adds functionality)

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)

[#3275](https://github.com/tikv/tikv/issues/3275)
The related source in tidb:
[https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4651](https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4662)
[https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4651](https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4679)
[https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4651](https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4705)
[https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4651](https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L4750)

### CODE REVIEW

**Comment by rleungx** [comment]
@Seact Thanks for your PR! Could you add some tests?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -322,6 +322,83 @@ impl ScalarFunc {
...
+        }
+    }
+
+    pub fn period_add(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let p = try_opt!(self.children[0].eval_int(ctx, row));
+        if p == 0 {
+            return Ok(Some(0));
+        }
+        let n = try_opt!(self.children[1].eval_int(ctx, row));
+        let (month, overflowing) = Self::period_to_month(u64::from_ne_bytes(p.to_ne_bytes()))
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: could we use `p as u64` directly?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -322,6 +322,83 @@ impl ScalarFunc {
         Ok(None)
     }
 
+    fn period_to_month(period: u64) -> u64 {
+        if period == 0 {
+            return 0;
+        }
+        let (year, month) = (period / 100, period % 100);
+        year * 12 + month - 1
+            + if year < 70 {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could you make it more easy to review?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 3]
```diff
@@ -322,6 +322,71 @@ impl ScalarFunc {
...
+        }
+    }
+
+    fn month_to_period(month: u64) -> u64 {
+        // maybe overflow
+        if month == 0 {
+            return 0;
+        }
+        let year = month / 12;
+        const FINAL_PERIOD: u64 = 206812;
```
Comment (crazycs520) [inline src/coprocessor/dag/expr/builtin_time.rs]: Sorry, I don't understand the`206812`.

I execute below SQL In MySQL:
```sql
mysql > select period_add(200212, 9223372036854775807)
+-----------------------------------------+
| period_add(200212, 9223372036854775807) |
+-----------------------------------------+
| 200211                                  |
+-----------------------------------------+
1 row in set
Time: 0.006s
mysql > select period_add(200212, 9223372036854775808)
+-----------------------------------------+
| period_add(200212, 9223372036854775808) |
+-----------------------------------------+
| 200212                                  |
+-----------------------------------------+
1 row in set
Time: 0.006s
mysql > select period_add(200212, 9223372036854775809)
+-----------------------------------------+
| period_add(200212, 9223372036854775809) |
+-----------------------------------------+
| 200301                                  |
+-----------------------------------------+
```

Comment (crazycs520) [reply src/coprocessor/dag/expr/builtin_time.rs]: Maybe we can refer to MySQL : https://github.com/mysql/mysql-server/blob/8.0/sql/item_timefunc.cc#L956

**Comment by rleungx** [comment]
@Seact CI failed. And please sign-off the DCO.

**Pull request closed without merging**